server/src/websocket.rs

```diff
@@ -0,0 +1,228 @@
+    Edit {
+        // Saving room for more fields here
+        source: String,
+    },
+    Compile,
+    Step,
+}
+
+/// All the different types of events that we can transmit over the websocket.
+/// This can include both success and error events.
+#[derive(Debug, Serialize)]
+#[serde(tag = "type", content = "content", rename_all = "snake_case")]
+enum OutgoingEvent<'a> {
+    // OK events
+    /// Send latest version of the program source code
+    Source {
+        // Saving room for more fields here
+        source: &'a str,
+    },
+    /// Send latest version of the machine state
+    MachineState {
+        state: &'a MachineState,
+        is_complete: bool,
+        is_successful: bool,
+    },
+
+    // Error events
+    /// Failed to parse websocket message
+    MalformedMessage(String),
+    /// Failed to parse the sent program
+    CompileError(CompileError),
+    /// Error occurred while running a program
+    RuntimeError(RuntimeError),
+    /// "Step" message occurred before "Compile" message
+    NoCompilation,
+}
+
+// Define type conversions to make processing code a bit cleaner
+
+impl<'a> From<&'a Machine> for OutgoingEvent<'a> {
+    fn from(other: &'a Machine) -> Self {
+        OutgoingEvent::MachineState {
+            state: other.get_state(),
+            is_complete: other.is_complete(),
+            is_successful: other.is_successful(),
+        }
+    }
+}
+
+impl<'a> From<serde_json::Error> for OutgoingEvent<'a> {
+    fn from(other: serde_json::Error) -> Self {
+        OutgoingEvent::MalformedMessage(format!("{}", other))
+    }
+}
+
+impl<'a> From<CompileError> for OutgoingEvent<'a> {
+    fn from(other: CompileError) -> Self {
+        OutgoingEvent::CompileError(other)
+    }
+}
+
+impl<'a> From<RuntimeError> for OutgoingEvent<'a> {
+    fn from(other: RuntimeError) -> Self {
+        OutgoingEvent::RuntimeError(other)
+    }
+}
+
+/// The controlling struct for a single websocket instance
+struct ProgramWebsocket {
+    /// Track the last time we pinged/ponged the client, if this exceeds
+    /// CLIENT_TIMEOUT, drop the connection
+    heartbeat: Instant,
+    /// The current user-entered program source code
+    source_code: String,
+    /// The current execution state of the machine. None if the program hasn't
+    /// been compiled yet.
+    machine: Option<Machine>,
+}
+
+impl ProgramWebsocket {
+    fn new() -> Self {
+        ProgramWebsocket {
+            heartbeat: Instant::now(),
+            source_code: String::new(),
+            machine: None,
+        }
+    }
+
+    /// Processes the given text message, and returns the appropriate response
+    /// event. The return type on this is a little funky because all our
+    /// event types (OK and error) are under the same enum. We still use a
+    /// Result because it makes it easier to exit early in the case of an error.
+    fn process_msg(
+        &mut self,
+        text: String,
+    ) -> Result<OutgoingEvent, OutgoingEvent> {
```

Comment:
im still confused on why its needs to be here twice

Reply:
It doesn't _need_ to be, but by returning a result, it makes it easier to handle errors. There are 3 spots in the function where we use the `?` operator, which can only be used when you're returning a `Result` (mostly). Conceptually from the outside it would make more sense to just return one `OutgoingEvent`, but if we did that, then the error-handling cases would have to look like
```rust
let socket_msg = match serde_json::from_str::<IncomingEvent>(&text) {
    Ok(msg) => msg,
    Err(err) => return err.into(),
};
```

and I didn't really want to write that in 3 different places.

Reply:
Im still a lil confused but i trust you lol